TestHelpers/struct_field_to_index_map_generator.py

```diff
@@ -0,0 +1,312 @@
+
+    print(var_decl_chunks)
+
+    if var_decl_chunks[5] == "used":
+        var_name = var_decl_chunks[6]
+    else:
+        var_name = var_decl_chunks[5]
+
+    struct_name = ""
+    start_index = var_decl.find("'") + 1
+    cur_char = var_decl[start_index]
+
+    while (cur_char != "'"):
+        struct_name += cur_char
+
+        start_index += 1
+        cur_char = var_decl[start_index]
+
+    struct_name_split = struct_name.split(" ")
+    if len(struct_name_split) > 1 and struct_name_split[1] != "":
+        struct_name = struct_name_split[1]
+    else:
+        struct_name = struct_name_split[0]
+
+    if not structs_manager.struct_exists(struct_name):
+        return
+
+    logout(var_decl)
+    logout(
+        f"parsed var_decl var name = {var_name} struct name = '{struct_name}'")
+
+    struct_var_manager.add_struct_var(var_name, struct_name)
+
+
+def add_typedef_alias(structs_manager: StructsManager, typedef_decl: str):
+    if not ("referenced" in typedef_decl and "struct" in typedef_decl):
+        return
+
+    original_name = ""
+    start_index = typedef_decl.find("'") + 1
+    cur_char = typedef_decl[start_index]
+
+    while (cur_char != "'"):
+        original_name += cur_char
+
+        start_index += 1
+        cur_char = typedef_decl[start_index]
+
+    original_name_split = original_name.split(" ")
+    if len(original_name_split) > 1 and original_name_split[1] != "":
+        original_name = original_name_split[1]
+    else:
+        original_name = original_name_split[0]
+
+    typedef_alias_name = ""
+    start_index = typedef_decl.find("referenced") + len("referenced") + 1
+    cur_char = typedef_decl[start_index]
+
+    while (cur_char != " "):
+        typedef_alias_name += cur_char
+
+        start_index += 1
+        cur_char = typedef_decl[start_index]
+
+    if original_name != typedef_alias_name:
+        structs_manager.get_struct(
+            original_name, False).add_typedef(typedef_alias_name)
+        logout(f"added typedef '{original_name}' '{typedef_alias_name}'")
+
+
+file_to_read = sys.argv[1]
+output_file = sys.argv[2]
+
+with open(file_to_read) as ast:
+    cur_struct = None
+    cur_top_decl = None  # FunctionDecl or RecordDecl (strings)
+    cur_mid_decl = None  # ParmVarDecl or FieldDecl (strings)
+    field_index = None
+
+    ast_lines = ast.readlines()
+    structs_manager = StructsManager()
+    struct_var_manager = StructVarManager()
+
+    for expr in ast_lines:
+        expr = expr.strip()
+
+        first_letter_index = expr.find(next(filter(str.isalpha, expr)))
+        start_decl_index = first_letter_index
+        decl = ""
+        cur_decl_char = expr[start_decl_index]
+
+        try:
+            while (cur_decl_char != " "):
+                decl += cur_decl_char
+                start_decl_index += 1
+                cur_decl_char = expr[start_decl_index]
+        except IndexError:
+            # might be a null stement (looks like "<<<NULL>>>" in AST)
```

Comment:
I'd prefer that we avoid exceptional control flow here. Can you check directly for `<<<NULL>>>` before you try to access the array instead of catching the `IndexError`?

(Exceptional control flow is both much slower than an `if` statement and tends to be harder for humans to reason about, in my experience.)

Reply:
It's hard for me to directly check for null statements without making too many assumptions about the clang AST string. It's also hard to clear up my confusions about it since the documentation on it is very poor (as far as I've searched). Thus, this exception is kind of a catch-all for any unfamiliar AST statements; if there are more unfamiliar AST statements I come across, I'll try to update this section to be more clear on what types of strange statements we're looking for

Reply:
Even if we want to keep the exceptional control flow for unexpected statements, for AST statements that we **know** cause this flow (at least `<<<NULL>>>`, possibly others), it would be better to add an explicit check.

I'm okay with leaving it as-is if you add a TODO indicating that the documentation is too poor for you to do better.